concepts/functions/about.md

```diff
@@ -1,3 +1,176 @@
+function contains no statements or expressions the curly braces are still required for the function definition. Of course there is no point to
+having an empty function.
+
+A function could contain a simple statement. In the following example the function takes in one `i32` argument and prints it
+
+```rust
+fn print_integer(value: i32) {
+    println!("{:?}", value);
+}
+```
+
+Note that the parameter needs to be given a name (`value` in this case) and be annotated with its type.
+
+A function can also return an expression. In the following example the function takes in one `i32` argument and returns it doubled
+
+```rust
+fn double_integer(value: i32) -> i32 {
+    value * 2
+}
+```
+
+The `-> i32` indicates that the function returns an `i32`. Unlike parameters, the returned value is not named. Note that the last expression
+in a function can be returned without the `return` keyword and a semicolon. `return` _is_ needed to return early from a function, like so
+
+```rust
+fn long_function() -> i32 {
+    let some_condition = false;
+    // code snipped
+    if (some_condition) {
+        return 0;
+    }
+    // more code snipped
+    42
+}
+```
+
+A function has its own scope which borrows or takes ownership of arguments passed to it. In this example the last line in `main` causes a
+compile error
+
+```rust
+fn double_integer(value: Box<i32>) -> i32 {
+    *value * 2
+}
+
+pub fn main() {
+    let num = Box::new(42);
+    let num_doubled = double_integer(num); // value moved here
+    println!("{:?}", num_doubled);
+    println!("{:?}", num); // borrow of moved value: `num` value borrowed here after move
+    // move occurs because `num` has type `Box<i32>`, which does not implement the `Copy` trait
+}
+```
+
+`main`, which owns the boxed `num` value, moves `num` as the argument for `value` into `double_integer`, which takes ownership of it. At the
+end of `double_integer`, `value` goes out of scope and is dropped, so `num` is no longer valid for the rest of the `main` function.
+
+Note that `fn main` is prefaced by `pub`, but `fn double_integer` is not. A `pub fn` is public, meaning it can be accessed from a different
+module. Functions defined without `pub` are private and can only be accessed from the current module. `pub(crate) fn` makes a function
+accessible only within the current crate.
+
+`const fn` is used to define a [constant function][constant function]. Going back to `double_integer`, by making it a `const fn` we can now use
+it to set a constant, like so
+
+```rust
+const fn double_integer(value: i32) -> i32 {
+    value * 2
+}
+
+pub fn main() {
+    const DOUBLE_42: i32 = double_integer(42);
+    println!("{:?}", DOUBLE_42);
+}
+```
+
+The constraint for a constant function is that it must be able to be evaluated at compile time. If the function does not satify that constraint
+a compile error will result when using `const fn`, as exampled below
```

Comment:
```suggestion
Because constant functions may be evaluated at compile time, they have some restrictions that normal functions do not.
In particular, a `const fn` can only call other functions also marked as `const`. 
Failure to abide by that restriction will result in a compile error.
```